### PR TITLE
Improvements & v 0.1

### PR DESCRIPTION
**Pull request by mmacai**
TODO:

- [x] cleanup
- [x] consolidate versions
- [x] add unit tests

### CODE REVIEW

File: lib/cloudevents/format/v_0_1/decoder/json.ex [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      iex> Cloudevents.Format.V_0_2.Decoder.JSON.decode(json)
+      {:error, %Cloudevents.Format.Decoder.DecodeError{
+                cause: %Cloudevents.Format.ParseError{message: "invalid extension attributes: \\"an extension attribute that contains spaces\\""}}}
+
+  """
+  def decode(json) do
+    with {:ok, orig_map} <- Jason.decode(json),
+         preprocessed_map = handle_data_base64(orig_map),
+         {:ok, event} <- Event.from_map(preprocessed_map) do
+      # `data_base64` may only be used for binary data, so we only try to interpret
```
Comment (kevinbader) [inline lib/cloudevents/format/v_0_1/decoder/json.ex]: I think in 0.1 base64-encoded data isn't even supported, or is it?

File: lib/cloudevents/format/v_0_1/encoder/map.ex [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    all_but_extensions =
+      %{
+        cloudEventsVersion: "0.1",
+        eventType: eventType,
+        source: source,
+        eventID: eventID
+      }
+      |> add_if_set(event, :subject)
+      |> add_if_set(event, :time)
+      # |> add_if_set(event, :datacontenttype)
```
Comment (kevinbader) [inline lib/cloudevents/format/v_0_1/encoder/map.ex]: `contentType`

File: lib/cloudevents/kafka_binding/v_1_0/decoder.ex [hunk 3]
```diff
@@ -15,16 +15,38 @@ defmodule Cloudevents.KafkaBinding.V_1_0.Decoder do
...
     case content_type(kafka_headers) do
-      "application/cloudevents" -> parse_structured(kafka_body, "json")
-      "application/cloudevents+" <> event_format -> parse_structured(kafka_body, event_format)
-      "application/cloudevents-batch" <> _ -> {:error, :batch_mode_not_available_as_per_spec}
-      event_format -> parse_binary(kafka_headers, kafka_body, event_format)
+      "application/cloudevents" ->
+        parse_structured(kafka_body, "json")
+
+      "application/cloudevents+avro" ->
+        parse_binary(kafka_headers, kafka_body, :avro)
```
Comment (kevinbader) [inline lib/cloudevents/kafka_binding/v_1_0/decoder.ex]: 🐛  `parse_binary` is for binary _mode_, just like `parse_structured` is for structured mode. The event format here might be Avro, which is a binary encoding; but the message itself is still _structured mode_.

File: lib/cloudevents/kafka_binding/v_1_0/decoder.ex [hunk 4]
```diff
@@ -61,8 +87,14 @@ defmodule Cloudevents.KafkaBinding.V_1_0.Decoder do
...
-    |> hd()
-    |> String.downcase()
+    content_type = for({"content-type", content_type} <- headers, do: content_type)
+
+    if length(content_type) == 1 do
+      content_type
+      |> hd()
+      |> String.downcase()
+    else
+      nil
```
Comment (kevinbader) [inline lib/cloudevents/kafka_binding/v_1_0/decoder.ex]: I think this shouldn't return `nil` but `application/cloudevents+json` instead, cause this is what we default to (the `case` expression above reads nicer then as well).

File: lib/cloudevents/format/v_0_1/decoder/json.ex (hunk 1 above)
Comment (mmacai) [reply lib/cloudevents/format/v_0_1/decoder/json.ex]: Seems like it's not.

File: lib/cloudevents/kafka_binding/v_1_0/decoder.ex (hunk 4 above)
Comment (mmacai) [reply lib/cloudevents/kafka_binding/v_1_0/decoder.ex]: Ye, makes sense.

File: lib/cloudevents/format/v_0_1/decoder/json.ex [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      {:error, %Cloudevents.Format.Decoder.DecodeError{
+                cause: %Cloudevents.Format.ParseError{message: "invalid extension attributes: \\"an extension attribute that contains spaces\\""}}}
+
+  """
+  def decode(json) do
+    with {:ok, orig_map} <- Jason.decode(json),
+         {:ok, event} <- Event.from_map(orig_map) do
+      data =
+        if Map.has_key?(orig_map, "data_base64") do
+          event.data
```
Comment (kevinbader) [inline lib/cloudevents/format/v_0_1/decoder/json.ex]: remove?

Comment (mmacai) [reply lib/cloudevents/format/v_0_1/decoder/json.ex]: Ye, have it locally, will be included in next push.

File: lib/cloudevents/kafka_binding/v_1_0/decoder.ex [hunk 6]
```diff
@@ -80,15 +77,18 @@ defmodule Cloudevents.KafkaBinding.V_1_0.Decoder do
...
     if length(content_type) == 1 do
       content_type
       |> hd()
       |> String.downcase()
     else
-      "application/cloudevents+json"
+      case kafka_body do
+        <<0::8, _id::32, _body::binary>> -> "avro/binary"
+        _ -> "application/cloudevents+json"
+      end
```
Comment (kevinbader) [inline lib/cloudevents/kafka_binding/v_1_0/decoder.ex:91]: I like this! Now all content-type related guesswork is at one place :)

**Comment by mmacai** [comment]
@kevinbader checked with RIG and functionally it looks alright. Please do another review so we can release it as soon as possible.

**Review by kevinbader — APPROVED** [review]
Looks good!

**Pull request merged**